README.md
# NX development by ArticaProject, X2Go and TheQVD

This source tree started as a re-distribution of those NX packages needed
to setup FreeNX and/or X2Go on a Linux server.

In the past, the NX re-distribution (3.5.0.x) had been maintained by the X2Go Project:
http://wiki.x2go.org

In 2014, [the QVD project](http://theqvd.com) run by the company Qindel
joined the group of people being interested in NX maintenance and
improvement.

Since 2015, the Arctica Project has joined in the NX development. The core devs
of X2Go, Arctica and TheQVD have agreed on stopping to redistribute NX and to
continue the development of NX 3.x as the new upstream instead. The package
will see a slight name change to **nx-libs** starting with version 3.5.99.0.

Our intentions for nx-libs are:

* provide _one_ tarball that builds NX projects via a common Makefile
* provide _one_ tarball for distribution packagers
* provide support for security issues
* provide support for latest X11 extensions
* improve NX where possible while still staying compatible to FreeNX and NoMachine 3.x


This source tree is maintained on Github:

  https://github.com/ArcticaProject/nx-libs (3.6.x branch)

For the the post-NoMachine era of nx-libs, we will focus on two release
phases for the upcoming two years.

## Release series 3.6.0.x

Release goals (phase 1) for nx-libs release series 3.6.0.x:

* CVE security audit (complete)
* remove unused code (+/- complete)
* no bundled non-X11 libraries anymore (complete)
* complete imake cleanup (+/- complete)
* replace as many libNX_X* libraries by X.org's libX* libraries
  (complete, only remaining library: libNX_X11)
* support for iOS (nxproxy, complete)
* Unix file socket communication for nxproxy -C <-> nxproxy -S connections
  (complete)
* allow Unix file sockets as channel endpoints (complete)
* allow embedding of nxproxy into other windows (work pending)
* new RandR based Xinerama extension (+/- complete, more QA needed)
* Fix Xcomposite extension in Xserver (work pending)
* nxcomp protocol clean-up (complete)
* nxcomp logging clean-up (work pending)
* optimizing documentation: how to tune NX connections (work pending)


## Release series 3.7.0.x

Release goals (phase 2) for nx-libs release series 3.7.0.x (not branched-off, yet):

* rebase Xserver code against latest X.Org server (work in progress)
* event FIFO sockets for attaching external applications
  (todo, to be discussed)
* enable/support XV extension (todo)
* software cursor for shadow sessions (todo)
* no bundled Mesa library anymore (todo, to be discussed)
* use recent MesaGL (todo, to-be-discussed)


If you have any questions about this NX development or want to file a bug
report, please contact the Arctica developers, the X2Go developers or the
TheQVD developers via the project's Github issue tracker.

*Sincerely,*

*The nx-libs developers (a combined effort of ArcticaProject / TheQVD / X2Go)*

## Building Under Fedora or EPEL using Mock

Assuming:

1. The branch you are building is 3.6.x
2. The current version is 3.5.99.0 (specified in the .spec file)
3. The current release is 0.0build1 (specified in the .spec file)
4. You wish for the RPM files and the mock build logs to be under ~/result

Prerequisites:

1. Install package "mock"
2. Add your user account to the "mock" group (recommended)
3. cd to the nx-libs directory that you cloned using git

```
mkdir -p ~/result
git archive -o ../nx-libs-3.5.99.0.tar.gz --prefix=nx-libs-3.5.99.0/ 3.6.x
cp --preserve=time nx-libs.spec ../
cd ..
mock --buildsrpm --spec ./nx-libs.spec --sources ./nx-libs-3.5.99.0.tar.gz --resultdir ~/result
mock --rebuild ~/result/nx-libs-3.5.99.0-0.0build1.fc23.src.rpm --resultdir ~/result
```

The end result is RPMs under ~/result that you can install (or upgrade to) using yum or dnf, which will resolve their dependencies.

## Building for openSUSE using OBS Build

Assuming:

1. The branch you are building is 3.6.x
2. The current version is 3.5.99.0 (specified in the .spec file)
3. The current release is 0.0build1 (specified in the .spec file)
4. You wish for the RPM files and the obs-build logs to be under ~/rpmbuild

Prerequisites:

1. Install package "obs-build"
2. Make sure your user account can become root via sudo
3. cd to the nx-libs directory that you cloned using git

```
mkdir -p ~/rpmbuild/SOURCES ~/rpmbuild/RPMS ~/rpmbuild/SRPMS ~/rpmbuild/OTHER ~/rpmbuild/BUILD
git archive -o $HOME/rpmbuild/SOURCES/nx-libs-3.5.99.0.tar.gz --prefix=nx-libs-3.5.99.0/ 3.6.x
cp --preserve=time nx-libs.spec ~/rpmbuild/SOURCES
cd ..
sudo obs-build --clean --nosignature --repo http://download.opensuse.org/distribution/<OPENSUSE-VERSION>/repo/oss/suse/ --root /var/lib/obs-build/ ~/rpmbuild/SOURCES/nx-libs.spec
cp -ar /var/lib/obs-build/home/abuild/rpmbuild/RPMS/* ~/rpmbuild/RPMS/
cp -ar /var/lib/obs-build/home/abuild/rpmbuild/SRPMS/* ~/rpmbuild/SRPMS/
cp -ar /var/lib/obs-build/home/abuild/rpmbuild/OTHER/* ~/rpmbuild/OTHER/
cp -ar /var/lib/obs-build/.build.log ~/rpmbuild/BUILD/
```

The end result is RPMs under ~/result that you can install (or upgrade to) using yum or dnf, which will resolve their dependencies.

## Building Under Debian or Ubuntu using debuild

Assuming:

1. The current version is 3.5.99\* (specified in debian/changelog)

Prerequisites:

1. Install package "devscripts"
2. Install the build dependencies. `dpkg-checkbuilddeps` can help you identify them.

```
git clone <url> nx-libs
cd nx-libs
debuild -uc -us
cd ..
sudo dpkg -i *3.5.99*.deb
```

## Building on Windows

The only components that can be built on Windows at the time of writing are `nxcomp` and `nxproxy` (with the latter utilizing the former).

**The next section is only relevant for git-based source code builds. Released tarballs do not require special handling.**

Since this project makes use of UNIX-style symlinks, it is imperative to clone the repository using Cygwin's git binary. MSYS(2) git is not able to handle UNIX-style symlinks.
Make sure that all build utilities are Cygwin-based. Non-Cygwin binaries will bail out with errors during the build process or insert garbage.


## Binary Builds

Most major Linux distribution come with nx. However, their packages
are often not the most current releases so we encourage you to update
to the newest release.

### Debian/Ubuntu
You can obtain binary builds of nx-libs for Debian and Ubuntu via these apt-URLs:

Debian: deb http://packages.arctica-project.org/debian {YOUR DIST VERSION} main
Ubuntu: deb http://packages.arctica-project.org/ubuntu {YOUR DIST VERSION} main

Our package server's archive key is: 0x98DE3101 (fingerprint: 7A49
CD37 EBAE 2501 B9B4 F7EA A868 0F55 98DE 3101). Use this command to
make APT trust our package server:

  wget -qO - http://packages.arctica-project.org/archive.key | sudo apt-key add -

### RedHat based

For RedHat based distributions current packages can be found at

https://bodhi.fedoraproject.org/updates/?packages=nx-libs

Please note that these are not always current as the builds are done manually.


## Compatibility with other NX based software

FreeNX is a free replacement of the original nxserver by
nomachine. OpenNX-CE is OSS GUI client replacing the original
nomachine nxclient. nx-libs is developed with those projects in mind
and tries to stay compatible with FreeNX and nomachine NX.  While
looking abandoned for many years some people are still using and
improving these packages. For current repos see

* https://github.com/dimbor-ru/freenx-server
* https://github.com/dimbor-ru/opennx

Thanks to dimbor for providing/maintaining these.

Etersoft is offering a commercial software called RX. Their OSS client
called rxclient also aims to be compatible to FreeNX and nomachine and
is based on opennx. You can find repos for their software here:
* https://github.com/Etersoft/rxclient
* https://github.com/Etersoft/rx-etersoft


ChangeLog
2023-06-28 Mike Gabriel

        * release 3.5.99.27 (HEAD -> 3.6.x, tag: 3.5.99.27)
        * Merge branch 'Ionic-bugfix/nxdialog-rpm-hashbang' into 3.6.x
          (cb929010f)

2023-06-27 Mihai Moldovan

        * nx-libs.spec: fix hashbang when manually replacing instead of using
          pathfix.py. (e925e37f0)

2023-04-28 Mike Gabriel

        * Merge branch 'uli42-pr/bump_years' into 3.6.x (4e70e0cb0)

2022-04-29 Ulrich Sibiller

        * Update Copyright years (390887b74)

2023-04-28 Mike Gabriel

        * Merge branch 'uli42-pr/silence_imake' into 3.6.x (70df7168c)

2022-01-28 Ulrich Sibiller

        * silence imake for normal builds (637a170e7)

2023-04-28 Mike Gabriel

        * Merge branch 'mjtrangoni-move-to-stream' into 3.6.x (fb0f8c1b7)

2022-03-28 Mario Trangoni

        * GHActions: Move to CentOS stream images (d6acd9253)

2023-04-27 Mike Gabriel

        * Merge branch 'uli42-pr/various9' into 3.6.x (9b94209f5)

2021-12-22 Ulrich Sibiller

        * Screen.c: use %s / __func__ for fprintf (16f6803f1)

2021-11-18 Ulrich Sibiller

        * Imakefiles: delete trailing whitespace (acd7de885)
        * nx-X11/config: whitespace fixes (f8a664e65)
        * testscripts: whitespace fixes (446a8ad69)
        * nxcomp: delete trailing whitespace (b40d0b1aa)
        * nxagent: delete trailing whitespace in many files (36c6f0943)
        * Loop.cpp: spelling fix (75c10f342)
        * Compext.c: spelling fixes (2d26be7da)
        * Compext.c: fix typo in datatype (9859a79d6)

2021-11-15 Ulrich Sibiller

        * Pointer.c: fix trailing whitespace (eaa532710)

2023-04-27 Mike Gabriel

        * Merge branch 'uli42-pr/update_libX11_2022' into 3.6.x (91d9218c0)

2022-01-19 Ulrich Sibiller

        * Update libNX_X11 to upstream's libX11-1.7.3.1-10-gd60ede78
          (4875a15ca)

2023-04-27 Mike Gabriel

        * Merge branch 'uli42-pr/extend_readme' into 3.6.x (a502149a8)

2021-12-29 Ulrich Sibiller

        * README.md: add references to maintained opennx, rxclient and freenx
          repos (091e06e47)

2023-02-23 Mihai Moldovan

        * nx-libs.spec: only depend upon pathfix.py for Fedora 27+ and RHEL
          7+. (288394c18)
        * nx-libs.spec: fix TIRPC usage on *SuSE 15.3+. (b536e9f91)
        * debian/rules: fix version detection on testing and unstable.
          (ec4f09add)
        * nx-libs.spec: add BR: upon pathfix.py. (f16808513)
        * nx-libs.spec: fix nxdialog hashbang selection, especially on *SuSE.
          (d35c5d011)

2022-01-29 Ulrich Sibiller

        * Merge pull request #1028 from uli42/pr/fix_clipboard_callbacks
          (8129c1f6b)
        * Merge pull request #1033 from uli42/pr/nxclient_geometry
          (106d6c8d4)
        * Merge pull request #1035 from gratuxri/ubuntu_jammy (ae9a6b2e4)

2022-01-16 gratuxri

        * Add Ubuntu 22.04 (1a1b5e9c2)

2021-12-29 Ulrich Sibiller

        * Screen.c: fix handling of AllScreens without a running window
          manager (c8553c70e)
          Fixes: ArcticaProject/nx-libs#1030

2021-11-10 Ulrich Sibiller

        * NXdispatch.c: use defined value CurrentTime instead of 0
          (a8d7b0c8c)

2021-11-15 Ulrich Sibiller

        * Clipboard.c: fix some comments (5ec4d7b8a)

2021-11-11 Ulrich Sibiller

        * Events.c: fix typo in comment (ce9782280)

2021-11-10 Ulrich Sibiller

        * Clipboard.c: use a TimeStamp datatype for lastTimeChanged as dix
          does (7b22b2d40)
        * Clipboard.c: fix typo in array name (a86957b91)
        * Clipboard.c: re-add callbacks (1d9ba187b)

2021-10-30 Mike Gabriel

        * Merge branch 'uli42-pr/various8' into 3.6.x (b06b6b1ef)

2021-10-28 Ulrich Sibiller

        * Clipboard.c: fix/reformat some comments (70fde4c74)

2021-10-25 Ulrich Sibiller

        * Screen.c, Window.c: use "None" everywhere for FullscreenWindow
          (5323e2cea)
        * Screen.c: Fix wrong format specifier (cf5bb486d)

2021-10-18 Ulrich Sibiller

        * Imakefiles: whitespace fixes (3ab28297c)
        * Args.c: remove superfluous SAFE_free() call (5a6928f86)
        * Args.c: fix wrong comment (4bdf03add)
        * Args.c: remove some empty lines (2e99e6234)

2021-06-08 Ulrich Sibiller

        * nxdialog: use os.linesep to make it work on other systems
          (e23efeb11)
        * nxdialog: rework comments (6090ab111)

2021-10-30 Mike Gabriel

        * Merge branch 'uli42-pr/extend_x2go_check' into 3.6.x (1ef24c379)

2021-10-18 Ulrich Sibiller

        * nxagent: extend x2go check (3d76ed45e)

2021-10-30 Mike Gabriel

        * Merge branch 'sunweaver-pr/debian-use-tirpc-in-ghactions' into
          3.6.x (c26b9b291)
        * .github/workflows/nx-libs.ym: Always build with -DUseTIRPC=YES.
          (00cc40a3b)

2021-10-18 Ulrich Sibiller

        * github actions: Use TIRPC for debian:sid (d207d02f1)

2021-10-30 Mike Gabriel

        * Merge branch 'AlexGhiti-int/alex/riscv64_support' into 3.6.x
          (2124e2fae)

2021-09-10 Alexandre Ghiti

        * nx-X11: Add support for riscv64 architecture (460562852)

2021-10-18 Mike Gabriel

        * Merge branch 'uli42-pr/many_clipboard_fixes' into 3.6.x (b7034724a)

2021-09-11 Mike Gabriel

        * debian/rules: Define -DUseTIRPC=1 for nx-X11 build on Debian and
          Ubuntu versions that already have libtirpc. (31057d00b)
        * debian/control: Add B-D: libtirpc-dev. (46b75d818)
        * debian/rules: Define -DUseTIRPC=1 for nx-X11 build on Debian and
          Ubuntu versions that already have libtirpc. (b67b910a6)
        * debian/control: Add B-D: libtirpc-dev. (6be4f87c6)

2021-06-21 Ulrich Sibiller

        * Clipboard.c: Add debug output when selection is already cleared
          (8264b03e8)

2021-06-20 Ulrich Sibiller

        * Clipboard.c: check for invalid index (987402309)
        * Events.c: add another comment (5da2be394)

2021-06-16 Ulrich Sibiller

        * Fix Xfixes event handling (7aa969cd4)

2021-06-22 Ulrich Sibiller


nxcomp/MD5.c
/*
  Copyright (C) 1999, 2000, 2002 Aladdin Enterprises.  All rights reserved.

  This software is provided 'as-is', without any express or implied
  warranty.  In no event will the authors be held liable for any damages
  arising from the use of this software.

  Permission is granted to anyone to use this software for any purpose,
  including commercial applications, and to alter it and redistribute it
  freely, subject to the following restrictions:

  1. The origin of this software must not be misrepresented; you must not
     claim that you wrote the original software. If you use this software
     in a product, an acknowledgment in the product documentation would be
     appreciated but is not required.
  2. Altered source versions must be plainly marked as such, and must not be
     misrepresented as being the original software.
  3. This notice may not be removed or altered from any source distribution.

  L. Peter Deutsch
  ghost@aladdin.com

 */
/* $Id: md5.c,v 1.6 2002/04/13 19:20:28 lpd Exp $ */
/*
  Independent implementation of MD5 (RFC 1321).

  This code implements the MD5 Algorithm defined in RFC 1321, whose
  text is available at
	http://www.ietf.org/rfc/rfc1321.txt
  The code is derived from the text of the RFC, including the test suite
  (section A.5) but excluding the rest of Appendix A.  It does not include
  any code or documentation that is identified in the RFC as being
  copyrighted.

  The original and principal author of md5.c is L. Peter Deutsch
  <ghost@aladdin.com>.  Other authors are noted in the change history
  that follows (in reverse chronological order):

  2002-04-13 lpd Clarified derivation from RFC 1321; now handles byte order
	either statically or dynamically; added missing #include <string.h>
	in library.
  2002-03-11 lpd Corrected argument list for main(), and added int return
	type, in test program and T value program.
  2002-02-21 lpd Added missing #include <stdio.h> in test program.
  2000-07-03 lpd Patched to eliminate warnings about "constant is
	unsigned in ANSI C, signed in traditional"; made test program
	self-checking.
  1999-11-04 lpd Edited comments slightly for automatic TOC extraction.
  1999-10-18 lpd Fixed typo in header comment (ansi2knr rather than md5).
  1999-05-03 lpd Original version.
 */

#include "MD5.h"

#include <string.h>

/*
 * Try to determine the CPU endianess
 * at compile time.
 */

#if defined(__linux) || defined(__CYGWIN32__)

#include <endian.h>

#if (__BYTE_ORDER == __LITTLE_ENDIAN)
#define ARCH_IS_BIG_ENDIAN 0
#else
#define ARCH_IS_BIG_ENDIAN 1
#endif

#endif /* #if defined(__linux) || defined(__CYGWIN32__) */

#undef BYTE_ORDER	/* 1 = big-endian, -1 = little-endian, 0 = unknown */
#ifdef ARCH_IS_BIG_ENDIAN
#  define BYTE_ORDER (ARCH_IS_BIG_ENDIAN ? 1 : -1)
#else
#  define BYTE_ORDER 0
#endif

#define T_MASK ((md5_word_t)~0)
#define T1 /* 0xd76aa478 */ (T_MASK ^ 0x28955b87)
#define T2 /* 0xe8c7b756 */ (T_MASK ^ 0x173848a9)
#define T3    0x242070db
#define T4 /* 0xc1bdceee */ (T_MASK ^ 0x3e423111)
#define T5 /* 0xf57c0faf */ (T_MASK ^ 0x0a83f050)
#define T6    0x4787c62a
#define T7 /* 0xa8304613 */ (T_MASK ^ 0x57cfb9ec)
#define T8 /* 0xfd469501 */ (T_MASK ^ 0x02b96afe)
#define T9    0x698098d8
#define T10 /* 0x8b44f7af */ (T_MASK ^ 0x74bb0850)
#define T11 /* 0xffff5bb1 */ (T_MASK ^ 0x0000a44e)
#define T12 /* 0x895cd7be */ (T_MASK ^ 0x76a32841)
#define T13    0x6b901122
#define T14 /* 0xfd987193 */ (T_MASK ^ 0x02678e6c)
#define T15 /* 0xa679438e */ (T_MASK ^ 0x5986bc71)
#define T16    0x49b40821
#define T17 /* 0xf61e2562 */ (T_MASK ^ 0x09e1da9d)
#define T18 /* 0xc040b340 */ (T_MASK ^ 0x3fbf4cbf)
#define T19    0x265e5a51
#define T20 /* 0xe9b6c7aa */ (T_MASK ^ 0x16493855)
#define T21 /* 0xd62f105d */ (T_MASK ^ 0x29d0efa2)
#define T22    0x02441453
#define T23 /* 0xd8a1e681 */ (T_MASK ^ 0x275e197e)
#define T24 /* 0xe7d3fbc8 */ (T_MASK ^ 0x182c0437)
#define T25    0x21e1cde6
#define T26 /* 0xc33707d6 */ (T_MASK ^ 0x3cc8f829)
#define T27 /* 0xf4d50d87 */ (T_MASK ^ 0x0b2af278)
#define T28    0x455a14ed
#define T29 /* 0xa9e3e905 */ (T_MASK ^ 0x561c16fa)
#define T30 /* 0xfcefa3f8 */ (T_MASK ^ 0x03105c07)
#define T31    0x676f02d9
#define T32 /* 0x8d2a4c8a */ (T_MASK ^ 0x72d5b375)
#define T33 /* 0xfffa3942 */ (T_MASK ^ 0x0005c6bd)
#define T34 /* 0x8771f681 */ (T_MASK ^ 0x788e097e)
#define T35    0x6d9d6122
#define T36 /* 0xfde5380c */ (T_MASK ^ 0x021ac7f3)
#define T37 /* 0xa4beea44 */ (T_MASK ^ 0x5b4115bb)
#define T38    0x4bdecfa9
#define T39 /* 0xf6bb4b60 */ (T_MASK ^ 0x0944b49f)
#define T40 /* 0xbebfbc70 */ (T_MASK ^ 0x4140438f)
#define T41    0x289b7ec6
#define T42 /* 0xeaa127fa */ (T_MASK ^ 0x155ed805)
#define T43 /* 0xd4ef3085 */ (T_MASK ^ 0x2b10cf7a)
#define T44    0x04881d05
#define T45 /* 0xd9d4d039 */ (T_MASK ^ 0x262b2fc6)
#define T46 /* 0xe6db99e5 */ (T_MASK ^ 0x1924661a)
#define T47    0x1fa27cf8
#define T48 /* 0xc4ac5665 */ (T_MASK ^ 0x3b53a99a)
#define T49 /* 0xf4292244 */ (T_MASK ^ 0x0bd6ddbb)
#define T50    0x432aff97
#define T51 /* 0xab9423a7 */ (T_MASK ^ 0x546bdc58)
#define T52 /* 0xfc93a039 */ (T_MASK ^ 0x036c5fc6)
#define T53    0x655b59c3
#define T54 /* 0x8f0ccc92 */ (T_MASK ^ 0x70f3336d)
#define T55 /* 0xffeff47d */ (T_MASK ^ 0x00100b82)
#define T56 /* 0x85845dd1 */ (T_MASK ^ 0x7a7ba22e)
#define T57    0x6fa87e4f
#define T58 /* 0xfe2ce6e0 */ (T_MASK ^ 0x01d3191f)
#define T59 /* 0xa3014314 */ (T_MASK ^ 0x5cfebceb)
#define T60    0x4e0811a1
#define T61 /* 0xf7537e82 */ (T_MASK ^ 0x08ac817d)
#define T62 /* 0xbd3af235 */ (T_MASK ^ 0x42c50dca)
#define T63    0x2ad7d2bb
#define T64 /* 0xeb86d391 */ (T_MASK ^ 0x14792c6e)


static void
md5_process(md5_state_t *pms, const md5_byte_t *data /*[64]*/)
{
    md5_word_t
	a = pms->abcd[0], b = pms->abcd[1],
	c = pms->abcd[2], d = pms->abcd[3];
    md5_word_t t;
#if BYTE_ORDER > 0
    /* Define storage only for big-endian CPUs. */
    md5_word_t X[16];
#else
    /* Define storage for little-endian or both types of CPUs. */
    md5_word_t xbuf[16];
    const md5_word_t *X;
#endif

    {
#if BYTE_ORDER == 0
	/*
	 * Determine dynamically whether this is a big-endian or
	 * little-endian machine, since we can use a more efficient
	 * algorithm on the latter.
	 */
	static const int w = 1;

	if (*((const md5_byte_t *)&w)) /* dynamic little-endian */
#endif
#if BYTE_ORDER <= 0		/* little-endian */
	{
	    /*
	     * On little-endian machines, we can process properly aligned
	     * data without copying it.
	     */
	    if (!((data - (const md5_byte_t *)0) & 3)) {
		/* data are properly aligned */
		X = (const md5_word_t *)data;
	    } else {
		/* not aligned */
		memcpy(xbuf, data, 64);
		X = xbuf;
	    }
	}
#endif
#if BYTE_ORDER == 0
	else			/* dynamic big-endian */
#endif
#if BYTE_ORDER >= 0		/* big-endian */
	{
	    /*
	     * On big-endian machines, we must arrange the bytes in the
	     * right order.
	     */
	    const md5_byte_t *xp = data;
	    int i;

#  if BYTE_ORDER == 0
	    X = xbuf;		/* (dynamic only) */
#  else
#    define xbuf X		/* (static only) */
#  endif
	    for (i = 0; i < 16; ++i, xp += 4)
		xbuf[i] = xp[0] + (xp[1] << 8) + (xp[2] << 16) + (xp[3] << 24);
	}
#endif
    }

#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32 - (n))))

    /* Round 1. */
    /* Let [abcd k s i] denote the operation
       a = b + ((a + F(b,c,d) + X[k] + T[i]) <<< s). */
#define F(x, y, z) (((x) & (y)) | (~(x) & (z)))
#define SET(a, b, c, d, k, s, Ti)\
  t = a + F(b,c,d) + X[k] + Ti;\
  a = ROTATE_LEFT(t, s) + b
    /* Do the following 16 operations. */
    SET(a, b, c, d,  0,  7,  T1);
    SET(d, a, b, c,  1, 12,  T2);
    SET(c, d, a, b,  2, 17,  T3);
    SET(b, c, d, a,  3, 22,  T4);
    SET(a, b, c, d,  4,  7,  T5);
    SET(d, a, b, c,  5, 12,  T6);
    SET(c, d, a, b,  6, 17,  T7);
    SET(b, c, d, a,  7, 22,  T8);
    SET(a, b, c, d,  8,  7,  T9);
    SET(d, a, b, c,  9, 12, T10);
    SET(c, d, a, b, 10, 17, T11);
    SET(b, c, d, a, 11, 22, T12);
    SET(a, b, c, d, 12,  7, T13);
    SET(d, a, b, c, 13, 12, T14);
    SET(c, d, a, b, 14, 17, T15);
    SET(b, c, d, a, 15, 22, T16);
#undef SET

     /* Round 2. */
     /* Let [abcd k s i] denote the operation
          a = b + ((a + G(b,c,d) + X[k] + T[i]) <<< s). */
#define G(x, y, z) (((x) & (z)) | ((y) & ~(z)))
#define SET(a, b, c, d, k, s, Ti)\
  t = a + G(b,c,d) + X[k] + Ti;\
  a = ROTATE_LEFT(t, s) + b
     /* Do the following 16 operations. */
    SET(a, b, c, d,  1,  5, T17);
    SET(d, a, b, c,  6,  9, T18);
    SET(c, d, a, b, 11, 14, T19);
    SET(b, c, d, a,  0, 20, T20);
    SET(a, b, c, d,  5,  5, T21);
    SET(d, a, b, c, 10,  9, T22);
    SET(c, d, a, b, 15, 14, T23);
    SET(b, c, d, a,  4, 20, T24);
    SET(a, b, c, d,  9,  5, T25);
    SET(d, a, b, c, 14,  9, T26);
    SET(c, d, a, b,  3, 14, T27);
    SET(b, c, d, a,  8, 20, T28);
    SET(a, b, c, d, 13,  5, T29);
    SET(d, a, b, c,  2,  9, T30);
    SET(c, d, a, b,  7, 14, T31);
    SET(b, c, d, a, 12, 20, T32);
#undef SET

     /* Round 3. */
     /* Let [abcd k s t] denote the operation
          a = b + ((a + H(b,c,d) + X[k] + T[i]) <<< s). */
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define SET(a, b, c, d, k, s, Ti)\
  t = a + H(b,c,d) + X[k] + Ti;\
  a = ROTATE_LEFT(t, s) + b
     /* Do the following 16 operations. */
    SET(a, b, c, d,  5,  4, T33);
    SET(d, a, b, c,  8, 11, T34);
    SET(c, d, a, b, 11, 16, T35);
    SET(b, c, d, a, 14, 23, T36);
    SET(a, b, c, d,  1,  4, T37);
    SET(d, a, b, c,  4, 11, T38);
    SET(c, d, a, b,  7, 16, T39);
    SET(b, c, d, a, 10, 23, T40);
    SET(a, b, c, d, 13,  4, T41);
    SET(d, a, b, c,  0, 11, T42);
    SET(c, d, a, b,  3, 16, T43);
    SET(b, c, d, a,  6, 23, T44);
    SET(a, b, c, d,  9,  4, T45);
    SET(d, a, b, c, 12, 11, T46);
    SET(c, d, a, b, 15, 16, T47);
    SET(b, c, d, a,  2, 23, T48);
#undef SET

     /* Round 4. */
     /* Let [abcd k s t] denote the operation
          a = b + ((a + I(b,c,d) + X[k] + T[i]) <<< s). */
#define I(x, y, z) ((y) ^ ((x) | ~(z)))
#define SET(a, b, c, d, k, s, Ti)\
  t = a + I(b,c,d) + X[k] + Ti;\
  a = ROTATE_LEFT(t, s) + b
     /* Do the following 16 operations. */
    SET(a, b, c, d,  0,  6, T49);
    SET(d, a, b, c,  7, 10, T50);
    SET(c, d, a, b, 14, 15, T51);
    SET(b, c, d, a,  5, 21, T52);
    SET(a, b, c, d, 12,  6, T53);
    SET(d, a, b, c,  3, 10, T54);
    SET(c, d, a, b, 10, 15, T55);
    SET(b, c, d, a,  1, 21, T56);
    SET(a, b, c, d,  8,  6, T57);
    SET(d, a, b, c, 15, 10, T58);
    SET(c, d, a, b,  6, 15, T59);
    SET(b, c, d, a, 13, 21, T60);
    SET(a, b, c, d,  4,  6, T61);
    SET(d, a, b, c, 11, 10, T62);
    SET(c, d, a, b,  2, 15, T63);
    SET(b, c, d, a,  9, 21, T64);
#undef SET

     /* Then perform the following additions. (That is increment each
        of the four registers by the value it had before this block
        was started.) */
    pms->abcd[0] += a;
    pms->abcd[1] += b;
    pms->abcd[2] += c;
    pms->abcd[3] += d;
}

void
md5_init(md5_state_t *pms)
{
    pms->count[0] = pms->count[1] = 0;
    pms->abcd[0] = 0x67452301;
    pms->abcd[1] = /*0xefcdab89*/ T_MASK ^ 0x10325476;
    pms->abcd[2] = /*0x98badcfe*/ T_MASK ^ 0x67452301;
    pms->abcd[3] = 0x10325476;
}

void
md5_append(md5_state_t *pms, const md5_byte_t *data, int nbytes)
{
    const md5_byte_t *p = data;
    int left = nbytes;
    int offset = (pms->count[0] >> 3) & 63;
    md5_word_t nbits = (md5_word_t)(nbytes << 3);

    if (nbytes <= 0)
	return;

    /* Update the message length. */
    pms->count[1] += nbytes >> 29;
    pms->count[0] += nbits;
    if (pms->count[0] < nbits)
	pms->count[1]++;

    /* Process an initial partial block. */
    if (offset) {
	int copy = (offset + nbytes > 64 ? 64 - offset : nbytes);

	memcpy(pms->buf + offset, p, copy);
	if (offset + copy < 64)
	    return;
	p += copy;
	left -= copy;
	md5_process(pms, pms->buf);
    }

    /* Process full blocks. */
    for (; left >= 64; p += 64, left -= 64)
	md5_process(pms, p);

    /* Process a final partial block. */
    if (left)
	memcpy(pms->buf, p, left);
}

void
md5_finish(md5_state_t *pms, md5_byte_t digest[16])
{
    static const md5_byte_t pad[64] = {
	0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
    };
    md5_byte_t data[8];
    int i;

    /* Save the length before padding. */
    for (i = 0; i < 8; ++i)
	data[i] = (md5_byte_t)(pms->count[i >> 2] >> ((i & 3) << 3));
    /* Pad to 56 bytes mod 64. */
    md5_append(pms, pad, ((55 - (pms->count[0] >> 3)) & 63) + 1);
    /* Append the length. */
    md5_append(pms, data, 8);
    for (i = 0; i < 16; ++i)
	digest[i] = (md5_byte_t)(pms->abcd[i >> 2] >> ((i & 3) << 3));
}
